server/src/main/java/org/apache/iotdb/db/engine/compaction/no/NoCompactionTsFileManagement.java

```diff
@@ -74,27 +86,48 @@ public NoCompactionTsFileManagement(String storageGroupName, String storageGroup
   @Override
   public void remove(TsFileResource tsFileResource, boolean sequence) {
     if (sequence) {
-      sequenceFileTreeSet.remove(tsFileResource);
+      synchronized (sequenceFileTreeSetMap) {
+        TreeSet<TsFileResource> sequenceFileTreeSet =
+            sequenceFileTreeSetMap.get(tsFileResource.getTimePartition());
+        sequenceFileTreeSet.remove(tsFileResource);
+      }
     } else {
-      unSequenceFileList.remove(tsFileResource);
+      synchronized (unSequenceFileListMap) {
+        List<TsFileResource> unSequenceFileList =
+            unSequenceFileListMap.get(tsFileResource.getTimePartition());
+        unSequenceFileList.remove(tsFileResource);
+      }
     }
   }
 
   @Override
   public void removeAll(List<TsFileResource> tsFileResourceList, boolean sequence) {
     if (sequence) {
-      sequenceFileTreeSet.removeAll(tsFileResourceList);
+      for (Set<TsFileResource> sequenceFileTreeSet : sequenceFileTreeSetMap.values()) {
+        sequenceFileTreeSet.removeAll(tsFileResourceList);
+      }
     } else {
-      unSequenceFileList.removeAll(tsFileResourceList);
+      for (List<TsFileResource> unSequenceFileList : unSequenceFileListMap.values()) {
+        unSequenceFileList.removeAll(tsFileResourceList);
+      }
```

Comment:
This contains potentially redundant operations, here is my advice:
1. sort `tsFileResourceList` according to their partitions
2. for each partition `p` existing in `tsFileResourceList`, find its first position `min` and last position `max` in `tsFileResourceList`
3. call `sequenceFileTreeSetMap.get(p).removeAll(tsFileResourceList.subList(min, max + 1))`

Reply:
By the way, is not concurrent safety a concern?